drivers/client_location_history/extensions/grda_warehouse/hmis_form_extension.rb

```diff
@@ -0,0 +1,79 @@
+###
+# Copyright 2016 - 2021 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module ClientLocationHistory::GrdaWarehouse
+  module HmisFormExtension
+    extend ActiveSupport::Concern
+
+    included do
+      has_many :client_locations, class_name: 'ClientLocationHistory::Location', as: :source
+
+      scope :with_location_data, -> do
+        joins(:hmis_assessment).merge(GrdaWarehouse::HMIS::Assessment.with_location_data)
+      end
+
+      def self.maintain_location_histories
+        ids = with_location_data.oldest_first.
+          where(
+            arel_table[:location_processed_at].eq(nil).
+            or(arel_table[:collected_at].gt(arel_table[:location_processed_at])),
+          ).pluck(:id)
+        return unless ids
+
+        # Remove any locations where the hmis_form no longer exists
+        location_source_ids = ClientLocationHistory::Location.where(source_type: 'GrdaWarehouse::HmisForm').
+          pluck(:source_id)
+        to_remove = location_source_ids - ids
```

Comment:
`ids` contains `HmisForms` that have not been processed, or have been updated since it was last processed, correct? Wouldn't this remove the locations for any forms that have been processed previously, but were not updated?

Reply:
Yes... we went round and round on this.  I'll try again.